s3/s3_fs.go

```diff
@@ -0,0 +1,235 @@
+package s3
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+	"github.com/spf13/afero"
+)
+
+// S3Fs is an FS object backed by S3.
+type S3Fs struct {
+	bucket string
+	s3API  s3iface.S3API
+}
+
+var _ afero.Fs = (*S3Fs)(nil)
+
+// NewS3Fs creates a new S3Fs object writing files to a given S3 bucket.
+func NewS3Fs(bucket string, s3API s3iface.S3API) *S3Fs {
+	return &S3Fs{bucket: bucket, s3API: s3API}
+}
+
+// Name returns the type of FS object this is: S3Fs.
+func (S3Fs) Name() string { return "S3Fs" }
+
+// Create a file.
+func (fs S3Fs) Create(name string) (afero.File, error) {
+	file, err := fs.Open(name)
+	if err != nil {
+		return file, err
+	}
+	// Create(), like all of S3, is eventually consistent.
+	// To protect against unexpected behavior, have this method
+	// wait until S3 reports the object exists.
+	if s3Client, ok := fs.s3API.(*s3.S3); ok {
+		return file, s3Client.WaitUntilObjectExists(&s3.HeadObjectInput{
+			Bucket: aws.String(fs.bucket),
+			Key:    aws.String(name),
+		})
+	}
+	return file, err
+}
+
+// Mkdir makes a directory in S3.
+func (fs S3Fs) Mkdir(name string, perm os.FileMode) error {
+	_, err := fs.OpenFile(fmt.Sprintf("%s/", filepath.Clean(name)), os.O_CREATE, perm)
+	return err
+}
+
+// MkdirAll creates a directory and all parent directories if necessary.
+func (fs S3Fs) MkdirAll(path string, perm os.FileMode) error {
+	return fs.Mkdir(path, perm)
+}
+
+// Open a file for reading.
+// If the file doesn't exist, Open will create the file.
+func (fs S3Fs) Open(name string) (afero.File, error) {
+	if _, err := fs.Stat(name); err != nil {
+		if os.IsNotExist(err) {
+			return fs.OpenFile(name, os.O_CREATE, 0777)
+		}
+		return (*S3File)(nil), err
```

Comment:
Any reason you're returning a valid interface to a nil pointer? While everyone should be checking the error, CacheOnReadFs checks the interface for nil, and this breaks in hard-to-troubleshoot ways. While I don't think anyone should be ignoring the error, returning a valid interface isn't doing anyone any favors either.